File: backend/app/schemas/question_bank.py
"""题库 Pydantic schema：分类 CRUD + 题目 CRUD + 组卷（assemble）请求/响应。

组卷（assemble）核心契约：
- 请求携带 ``quotas``（category_id → 目标分），省略时用各 category.target_points
- 响应返回选中题目 + 实际总分 + 各分类缺口（deficits）
- 凑不满 100 时由调用方决定：abort / 接受缺口 / AI 兜底
"""
from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.interview import InterviewDimensionLiteral

# ============================================================================
# 分类（QuestionCategory）
# ============================================================================


class CategoryBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=1, max_length=64)
    slug: str = Field(..., min_length=1, max_length=64, pattern=r"^[a-z0-9_-]+$")
    target_points: int = Field(default=0, ge=0, le=100)
    sort_order: int = Field(default=0, ge=0)


class CategoryCreate(CategoryBase):
    pass


class CategoryUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=64)
    target_points: int | None = Field(default=None, ge=0, le=100)
    sort_order: int | None = Field(default=None, ge=0)
    is_active: bool | None = None


class CategoryOut(CategoryBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    team_id: uuid.UUID
    is_active: bool


# ============================================================================
# 题目（QuestionBankItem）
# ============================================================================


class ItemBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    category_id: uuid.UUID
    dimension: InterviewDimensionLiteral | None = None
    question: str = Field(..., min_length=4, max_length=1000)
    points: int = Field(..., ge=1, le=100)
    difficulty: int | None = Field(default=None, ge=1, le=5)
    tags: list[str] | None = None
    reference_answer: str | None = Field(default=None, max_length=2000)

    @field_validator("question")
    @classmethod
    def _question_non_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("question cannot be empty")
        return v


class ItemCreate(ItemBase):
    pass


class ItemUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    category_id: uuid.UUID | None = None
    dimension: InterviewDimensionLiteral | None = None
    question: str | None = Field(default=None, min_length=4, max_length=1000)
    points: int | None = Field(default=None, ge=1, le=100)
    difficulty: int | None = Field(default=None, ge=1, le=5)
    tags: list[str] | None = None
    reference_answer: str | None = Field(default=None, max_length=2000)
    is_active: bool | None = None


class ItemOut(ItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    team_id: uuid.UUID
    is_active: bool


# ============================================================================
# 组卷（assemble）
# ============================================================================


class SignalInfo(BaseModel):
    """动态匹配信号（简历/JD 提取的技能词）。"""

    signal: str
    weight: float


class QuotaPlanItem(BaseModel):
    """某分类的动态配额计划项。"""

    category_id: uuid.UUID
    category_name: str
    base_points: int
    quota_points: int
    score: float
    matched: bool


class AssemblePlan(BaseModel):
    """动态组卷计划：信号 + 各分类配额调整（静态组卷时为 None）。"""

    total_target: int
    signals: list[SignalInfo]
    quotas: list[QuotaPlanItem]


class AssembleRequest(BaseModel):
    """按分类配额凑分组卷请求。

    ``quotas`` 省略时：动态模式用简历/JD 匹配后的动态配额；否则用各 category.target_points。
    ``tolerance`` 允许的分数容差（默认 ±5）。
    ``dynamic=true`` 需携带 ``session_id``（从中解析候选人/职位提取信号）。
    """

    model_config = ConfigDict(extra="forbid")

    quotas: dict[uuid.UUID, int] | None = Field(
        default=None, description="category_id → 目标分；省略则动态计算或用各分类 target_points"
    )
    tolerance: int = Field(default=5, ge=0, le=20)
    exclude_question_ids: list[uuid.UUID] | None = Field(
        default=None, description="组卷时排除的题目 id（避免对同一候选人重复出题）"
    )
    dynamic: bool = Field(
        default=False, description="按候选人简历 + JD 动态匹配配额（需 session_id）"
    )
    session_id: uuid.UUID | None = Field(
        default=None, description="面试会话 id（dynamic=true 时必填，用于提取信号）"
    )


class CategoryDeficit(BaseModel):
    """某分类凑分缺口。"""

    category_id: uuid.UUID
    category_name: str
    target: int
    actual: int
    gap: int


class AssembleResponse(BaseModel):
    """组卷结果：选中题目 + 实际总分 + 各分类缺口 + 动态配额计划。"""

    items: list[ItemOut]
    actual_total: int
    target_total: int
    deficits: list[CategoryDeficit]
    plan: AssemblePlan | None = None


# ============================================================================
# 实例化（instantiate：把选中题目写入 interview_questions）
# ============================================================================


ComposeSource = Literal["bank", "ai_fallback"]
"""组卷来源：题库选题 / AI 兜底补充。"""


__all__ = [
    "AssemblePlan",
    "AssembleRequest",
    "AssembleResponse",
    "CategoryBase",
    "CategoryCreate",
    "CategoryDeficit",
    "CategoryUpdate",
    "CategoryOut",
    "ItemBase",
    "ItemCreate",
    "ItemUpdate",
    "ItemOut",
    "QuotaPlanItem",
    "SignalInfo",
    "ComposeSource",
]

File: frontend/lib/api/questionBank.ts
"use client";

import { apiClient } from "./client";
import type { InterviewDimension } from "./interview";

/**
 * 题库 API client：分类 CRUD + 题目 CRUD + 按分类配额凑 100 分组卷。
 */

// ============================================================================
// 分类
// ============================================================================

export interface QuestionCategory {
  id: string;
  team_id: string;
  name: string;
  slug: string;
  target_points: number;
  sort_order: number;
  is_active: boolean;
}

export interface CategoryCreate {
  name: string;
  slug: string;
  target_points: number;
  sort_order?: number;
}

export interface CategoryUpdate {
  name?: string;
  target_points?: number;
  sort_order?: number;
  is_active?: boolean;
}

export async function listCategoriesApi(): Promise<QuestionCategory[]> {
  const { data } = await apiClient.get<QuestionCategory[]>(
    "/api/question-bank/categories",
  );
  return data;
}

export async function createCategoryApi(
  payload: CategoryCreate,
): Promise<QuestionCategory> {
  const { data } = await apiClient.post<QuestionCategory>(
    "/api/question-bank/categories",
    payload,
  );
  return data;
}

export async function updateCategoryApi(
  categoryId: string,
  payload: CategoryUpdate,
): Promise<QuestionCategory> {
  const { data } = await apiClient.patch<QuestionCategory>(
    `/api/question-bank/categories/${categoryId}`,
    payload,
  );
  return data;
}

export async function deleteCategoryApi(categoryId: string): Promise<void> {
  await apiClient.delete(`/api/question-bank/categories/${categoryId}`);
}

// ============================================================================
// 题目
// ============================================================================

export interface QuestionBankItem {
  id: string;
  team_id: string;
  category_id: string;
  dimension: InterviewDimension | null;
  question: string;
  points: number;
  difficulty: number | null;
  tags: string[] | null;
  reference_answer: string | null;
  is_active: boolean;
}

export interface ItemCreate {
  category_id: string;
  dimension?: InterviewDimension | null;
  question: string;
  points: number;
  difficulty?: number | null;
  tags?: string[] | null;
  reference_answer?: string | null;
}

export type ItemUpdate = Partial<ItemCreate> & { is_active?: boolean };

export async function listItemsByCategoryApi(
  categoryId: string,
): Promise<QuestionBankItem[]> {
  const { data } = await apiClient.get<QuestionBankItem[]>(
    `/api/question-bank/categories/${categoryId}/items`,
  );
  return data;
}

export async function createItemApi(payload: ItemCreate): Promise<QuestionBankItem> {
  const { data } = await apiClient.post<QuestionBankItem>(
    "/api/question-bank/items",
    payload,
  );
  return data;
}

export async function updateItemApi(
  itemId: string,
  payload: ItemUpdate,
): Promise<QuestionBankItem> {
  const { data } = await apiClient.patch<QuestionBankItem>(
    `/api/question-bank/items/${itemId}`,
    payload,
  );
  return data;
}

export async function deleteItemApi(itemId: string): Promise<void> {
  await apiClient.delete(`/api/question-bank/items/${itemId}`);
}

// ============================================================================
// 组卷（assemble / compose）
// ============================================================================

export interface AssembleRequest {
  quotas?: Record<string, number> | null;
  tolerance?: number;
  exclude_question_ids?: string[] | null;
  /** 按候选人简历 + JD 动态匹配配额（需 session_id） */
  dynamic?: boolean;
  session_id?: string | null;
}

export interface CategoryDeficit {
  category_id: string;
  category_name: string;
  target: number;
  actual: number;
  gap: number;
}

export interface SignalInfo {
  signal: string;
  weight: number;
}

export interface QuotaPlanItem {
  category_id: string;
  category_name: string;
  base_points: number;
  quota_points: number;
  score: number;
  matched: boolean;
}

export interface AssemblePlan {
  total_target: number;
  signals: SignalInfo[];
  quotas: QuotaPlanItem[];
}

export interface AssembleResponse {
  items: QuestionBankItem[];
  actual_total: number;
  target_total: number;
  deficits: CategoryDeficit[];
  plan?: AssemblePlan | null;
}

/** 预览组卷（不落库）：返回选中题 + 实际总分 + 各分类缺口。 */
export async function assembleApi(payload: AssembleRequest): Promise<AssembleResponse> {
  const { data } = await apiClient.post<AssembleResponse>(
    "/api/question-bank/assemble",
    payload,
  );
  return data;
}

export interface ComposeResponse {
  batch_id: string;
  question_count: number;
  actual_total: number;
  target_total: number;
  deficits: CategoryDeficit[];
  plan?: AssemblePlan | null;
}

/** 组卷并写入指定 session（新 batch_id）。 */
export async function composeFromBankApi(
  sessionId: string,
  payload: AssembleRequest,
): Promise<ComposeResponse> {
  const { data } = await apiClient.post<ComposeResponse>(
    `/api/interview/sessions/${sessionId}/compose`,
    payload,
  );
  return data;
}
